### PR TITLE
N-ary lambdas, and indexes as lambda parameters

### PR DESCRIPTION
**Pull request by taniabogatsch**
Before this PR, our lambda functions must always be unary, i.e., the list of lambda parameters is just a single parameter: `x -> expr`. This PR reworks the lambda binding functionality to allow for n-ary lambda parameters. This change lays the groundwork to support `list_reduce` in a future PR.

None of our lambda functions (`list_transform`, `list_filter`) require more than a single parameter. However, allowing n-ary lambdas enables supporting an index parameter to our existing lambda functions. This index parameter must always be the second parameter, containing the 1-based index of the current element in the list.

```sql
SELECT list_filter([1, 2, 1], (x, y) -> x >= y);
----
[1, 2]

SELECT list_transform([1, 2, 3], (x, x_i) -> list_transform([4, 5, 6], (y, y_i) -> x + y + x_i + y_i));
----
[[7, 9, 11], [9, 11, 13], [11, 13, 15]]
```

### Scalar function callback

To correctly assign the types of the different lambda parameters, we need to support callbacks to the specific scalar functions. Only these functions know precisely how their parameters will look (e.g., the first parameter is the list child, and the second is `UBIGINT` for indexes).

### Additional Notes

- I started cleaning up the lambda code and separated functionality for `list_transform` and `list_filter`. This is only an initial refacto,r and we should pick this up in a future PR (more code separation, no more flatten, more const).
- I added a `LAMBDA_REF` expression to better distinguish between column and lambda references during binding.

### CODE REVIEW

File: src/core_functions/scalar/list/list_transform.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+	switch (parameter_idx) {
+	case 0:
+		return list_child_type;
+	case 1:
+		return LogicalType::BIGINT;
+	default:
+		throw BinderException("This lambda function only supports up to two lambda parameters!");
+	}
+}
```
Comment (maiadegraaf) [inline src/core_functions/scalar/list/list_transform.cpp:36]: Could we turn this into a generic function, like `BindBinaryLambda`.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - one minor comment:

**Comment by taniabogatsch** [comment]
Note to myself: these two branches have pending PRs with bug fixes after this major PR is in.
- https://github.com/maiadegraaf/duckdb/tree/random_lists
- https://github.com/taniabogatsch/duckdb/tree/lambda-ser

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**